src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunTypeSystemHelpers.cs

```diff
@@ -0,0 +1,386 @@
+                            isExactMethodResolved = (method != null);
+                        }
+                    }
+
+                    if (!isExactMethodResolved)
+                    {
+                        // We couldn't resolve the interface statically
+                        // Notify the caller that it should use runtime lookup
+                        // Note that we can leave pMD incorrect, because we will use runtime lookup
+                        forceRuntimeLookup = true;
+                    }
+                }
+                else
+                {
+                    // If we can resolve the interface exactly then do so (e.g. when doing the exact 
+                    // lookup at runtime, or when not sharing generic code).
+                    if (constrainedCanonType.CanCastTo(interfaceType))
+                    {
+                        method = constrainedCanonType.GetMethodDescForInterfaceMethod((MetadataType)interfaceType, genInterfaceMethod, throwOnConflict: false);
+                        if (method == null)
+                        {
+                            Debug.WriteLine("TryResolveConstraintMethodApprox: failed to find method desc for interface method\n");
+                        }
+                    }
+                }
+            }
+            else if (genInterfaceMethod.IsVirtual)
+            {
+                method = constrainedType.FindVirtualFunctionTargetMethodOnObjectType(genInterfaceMethod);
+            }
+            else
+            {
+                // The method will be null if calling a non-virtual instance 
+                // methods on System.Object, i.e. when these are used as a constraint.
+                method = null;
+            }
+
+            if (method == null)
+            {
+                // Fall back to VSD
+                return null;
+            }
+
+            //#TryResolveConstraintMethodApprox_DoNotReturnParentMethod
+            // Only return a method if the value type itself declares the method, 
+            // otherwise we might get a method from Object or System.ValueType
+            if (!method.OwningType.IsValueType)
+            {
+                // Fall back to VSD
+                return null;
+            }
+
+            // We've resolved the method, ignoring its generic method arguments
+            // If the method is a generic method then go and get the instantiated descriptor
+            if (interfaceMethod.HasInstantiation)
+            {
+                method = method.MakeInstantiatedMethod(interfaceMethod.Instantiation);
+            }
+
+            Debug.Assert(method != null);
+            //assert(!pMD->IsUnboxingStub());
+
+            return method;
+        }
+
+        private static MethodDesc GetMethodDescForInterfaceMethod(this DefType thisType, MetadataType interfaceType, MethodDesc interfaceMethod, bool throwOnConflict)
```

Comment:
The fact that starting from this line, this file is forking the entire interface method resolution logic that we already have in the type system is concerning from a maintainability perspective. We have unit tests for the resolution provided by the type system, but we don't have anything for this. This new implementation is only used in the rare case of constrained calls. It will be a bug farm.

What problems were you hitting with the existing implementation in the type system?

This implementation is not correct. I don't think the `TryResolveConstraintMethodApprox` rewrite should be included in a pull request that just wants to rewrite crossgens getCallInfo in managed code. It should be a separate pull request.

The one and only purpose of `TryResolveConstraintMethodApprox` is to figure out that in code like this...

```csharp
using System;
using System.Runtime.CompilerServices;

class Program
{
    interface IFoo<T>
    {
        void Frob();
    }

    struct Foo<T> : /*IFoo<object>,*/ IFoo<string>
    {
        public int Value;

        //void IFoo<object>.Frob() => Value += 1;
        void IFoo<string>.Frob() => Value += 2;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Frob<T, U>(ref T x) where T : IFoo<U>
    {
        x.Frob();
    }

    static int Main()
    {
        //var a = new Foo<object>();
        //Frob<Foo<object>, object>(ref a);
        //Console.WriteLine(a.Value);
        var b = new Foo<string>();
        Frob<Foo<string>, string>(ref b);
        Console.WriteLine(b.Value);
        return 100;
    }
}
```

...we can resolve the destination method at compile time. The algorithm this is adding will not solve that and `TryResolveConstraintMethodApprox` still returns null. RyuJIT will then incorrectly attempt to box the struct. We only get saved by the fact that BOX_THIS then disqualifies the method from R2R and we don't end up generating the bad code.

Reply:
Hmm, so are you saying I'm free to make any changes to the existing TryResolveConstraintMethodApprox in the existing code shared between CPAOT and RyuJit? That would be certainly ideal, I was just under the impression that tons of minute details regarding treatment of shared generics in the compiler differ between CoreRT and CoreCLR and render such unification difficult.

Reply:
I don't see why we shouldn't be able to share the implementation.

Fixing `ResolveContraintMethodApprox` is definitely a P3 task. It's complicated logic that is only hit in very rare cases. The stubbed out version we currently use for full AOT never hit an issue in real world code. I remember David saying that even in the CLR, this came in as a bugfix pretty late because it's such a corner case. There are things with a better cost/benefit payoff right now.